File: client/platform/desktop/backend/serializers/kpf.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * DIVA KPF parser
 *
 * This parser is an example for how to turn a domain-specific annotation schema
 * like DIVA KPF into DIVE Json.
 */

import yaml from 'js-yaml';
import fs from 'fs-extra';
import { AnnotationFileData } from 'platform/desktop/backend/serializers/viame';
import { Feature } from 'vue-media-annotator/track';
import { ConfidencePair, StringKeyObject } from 'vue-media-annotator/BaseAnnotation';
import { MultiGroupRecord, MultiTrackRecord } from 'dive-common/apispec';
import { flattenDeep, uniqBy } from 'lodash';
import { GroupMembers } from 'vue-media-annotator/Group';

export const KPF_EXTENSIONS = ['activities.yml', 'geom.yml', 'types.yml'];

const GEOM_ACTOR_KEY = 'id1';
const TIME_FRAMES_KEY = 'ts0';
const GEOM_BBOX_KEY = 'g0';

interface KPFType {
  cset3: Record<string, number>;
  id1: number;
}

interface KPFGeom {
  g0: string;
  id0: number;
  id1: number;
  keyframe: boolean;
  ts0: number;
}

interface KPFTimespan {
  tsr0: [number, number];
}

interface KPFActor {
  id1: number;
  timespan: KPFTimespan[];
}

interface KPFActivity {
  id2: number;
  src_status: string;
  act2: Record<string, number>;
  actors: KPFActor[];
  timespan: KPFTimespan[];
}

/**
* KPF Class parses annotations defined in KPF yaml format.
*/
export default class KPF {
  activities: KPFActivity[];

  geom: KPFGeom[];

  types: KPFType[];

  actorType: Record<string, KPFType>;

  meta: Record<string, any>;

  actorGeom: Record<string, KPFGeom[]>;

  actorActivity: Record<string, KPFActivity[]>;

  constructor(
    activities: KPFActivity[] = [],
    geom: KPFGeom[] = [],
    types: KPFType[] = [],
    meta = {},
  ) {
    this.activities = activities;
    this.geom = geom;
    this.types = types;
    this.actorType = {};
    this.actorGeom = {};
    this.actorActivity = {};
    this.meta = meta;
    this.postProcess();
  }

  postProcess() {
    this.types = KPF.filterEmpty(this.types, [GEOM_ACTOR_KEY]);
    this.geom = KPF.filterEmpty(this.geom, [GEOM_ACTOR_KEY, GEOM_BBOX_KEY, TIME_FRAMES_KEY]);
    this.activities = KPF.filterEmpty(this.activities, ['act2']);

    this.types.forEach((t) => {
      const actorId = t[GEOM_ACTOR_KEY];
      this.actorType[actorId] = t;
    });

    this.geom.forEach((g) => {
      const actorId = g[GEOM_ACTOR_KEY];
      if (actorId in this.actorGeom) {
        this.actorGeom[actorId].push(g);
      } else {
        this.actorGeom[actorId] = [g];
      }
    });

    this.activities.forEach((act) => {
      act.actors.forEach((actor) => {
        const actorId = actor.id1;
        if (actorId in this.actorActivity) {
          this.actorActivity[actorId].push(act);
        } else {
          this.actorActivity[actorId] = [act];
        }
      });
    });
  }

  /**
  * Turn KPF into DIVE Json
  * @param {Object} staticMeta static metadata to assign to all tracks.
  * @returns {Array<tdm.Track>}
  */
  getJson(keyframesOnly = true): AnnotationFileData {
    const tracks: MultiTrackRecord = {};
    const groups: MultiGroupRecord = {};

    Object.keys(this.actorGeom).forEach((actorIdStr) => {
      const activity = this.actorActivity[actorIdStr];
      const actorType = this.actorType[actorIdStr];
      const attributes: StringKeyObject = {
        activityIds: activity.map((a) => a.id2).join(' '),
        srcStatus: activity.map((a) => a.src_status).join(' '),
      };
      const activityActor = flattenDeep(
        activity.map((v) => v.actors.filter((a) => a.id1 === actorType.id1)),
      );
      const allRanges = flattenDeep(
        activityActor.map((v) => v.timespan.map((ts) => ts.tsr0)),
      );
      const begin = Math.min(...allRanges);
      const end = Math.max(...allRanges);
      const features: Feature[] = [];
      const confidencePairs: ConfidencePair[] = Object.entries(actorType.cset3);
      Object.values(this.actorGeom[actorIdStr]).forEach((geom) => {
        if ((geom.keyframe && keyframesOnly) || !keyframesOnly) {
          features.push({
            frame: geom.ts0,
            attributes: {},
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            bounds: geom.g0.split(' ').map((f: string) => parseFloat(f)),
            interpolate: true,
          });
        }
      });

      tracks[actorType.id1] = {
        attributes,
        id: actorType.id1,
        begin,
        end,
        confidencePairs,
        features: uniqBy(features.sort((a, b) => a.frame - b.frame), 'frame'),
      };
    });

    this.activities.forEach((activity) => {
      const allRanges = flattenDeep(activity.timespan.map((v) => v.tsr0));
      const confidencePairs: ConfidencePair[] = Object.entries(activity.act2);
      const begin = Math.min(...allRanges);
      const end = Math.max(...allRanges);
      const members: GroupMembers = {};
      activity.actors.forEach((actor) => {
        members[actor.id1] = {
          ranges: actor.timespan.map((t) => t.tsr0),
        };
      });
      groups[activity.id2] = {
        begin,
        end,
        members,
        id: activity.id2,
        confidencePairs,
        attributes: {},
      };
    });

    return {
      tracks,
      groups,
    };
  }

  /**
  * Return elements of list that have validKey defined
  * @param {Array} list
  * @param {String} validKey
  */
  static filterEmpty(list: any[], validKeyList: string[]) {
    return list.filter((l) => validKeyList.some((e) => e in l));
  }

  /**
  * Return elements where number key in range value
  */
  static filterRange(list: any[], key: string, lower: number, upper: number) {
    return list.filter((item) => item[key] <= upper && item[key] >= lower);
  }

  /**
  * Create KPF object from YAML text strings.
  * @param {String} activityText
  * @param {String} geometryText
  */
  static fromText(activityText: string, geometryText: string, typeText: string) {
    const activityJson = (yaml.load(activityText) as any[]).map((a) => ({ ...a.act }));
    const geometryJson = (yaml.load(geometryText) as any[]).map((g) => ({ ...g.geom }));
    const typeJson = (yaml.load(typeText) as any[]).map((t) => ({ ...t.types }));
    return new KPF(activityJson, geometryJson, typeJson);
  }

  static async parse(
    activityFile: string, geometryFile: string, typeFile: string,
  ) {
    const promiseList = [];
    promiseList.push(fs.readFile(geometryFile, 'utf8'));
    promiseList.push(fs.readFile(typeFile, 'utf8'));
    promiseList.push(fs.readFile(activityFile, 'utf8'));
    const [geometryText, typeText, activityText] = await Promise.all(promiseList);
    const kpf = KPF.fromText(activityText, geometryText, typeText);
    return kpf.getJson();
  }
}

File: server/dive_utils/serializers/kpf.py
"""
This is a pure python module for parsing MEVA KPF.  It has no third party dependencies
and does not import code from other places in this application.

It is the python implementation of client/platform/backend/serializers/kpf.ts
"""
from collections import defaultdict
from typing import Dict, List, Mapping, Tuple, TypedDict, cast

import yaml

from dive_utils import constants, models, types


class KPFType(TypedDict):
    cset3: Mapping[str, float]
    id1: int


class KPFGeom(TypedDict):
    g0: str
    id0: int
    id1: int
    keyframe: bool
    ts0: int  # frame number


class KPFTimespan(TypedDict):
    tsr0: List[int]


class KPFActor(TypedDict):
    id1: int
    timespan: List[KPFTimespan]


class KPFActivity(TypedDict):
    id2: int
    src_status: str
    act2: Mapping[str, int]
    actors: List[KPFActor]
    timespan: List[KPFTimespan]


class KPFData(TypedDict):
    types: List[KPFType]
    geom: List[KPFGeom]
    activities: List[KPFActivity]
    actor_type_map: Dict[int, KPFType]
    actor_activity_map: Dict[int, List[KPFActivity]]
    actor_geom_map: Dict[int, List[KPFGeom]]


def flatten(container):
    for i in container:
        if isinstance(i, (list, tuple)):
            for j in flatten(i):
                yield j
        else:
            yield i


def get_default_kpf_data() -> KPFData:
    return {
        'types': [],
        'geom': [],
        'activities': [],
        'actor_type_map': {},
        'actor_activity_map': defaultdict(list),
        'actor_geom_map': defaultdict(list),
    }


def load(input: str) -> KPFData:
    """Load from files into formal data structures"""
    kpf_data = get_default_kpf_data()
    parsed = yaml.safe_load(input)
    for row in parsed:
        if 'types' in row:
            kpf_data['types'].append(cast(KPFType, row['types']))
        elif 'geom' in row:
            kpf_data['geom'].append(cast(KPFGeom, row['geom']))
        elif 'act' in row:
            kpf_data['activities'].append(cast(KPFActivity, row['act']))

    for typeval in kpf_data['types']:
        kpf_data['actor_type_map'][typeval['id1']] = typeval
    for geomval in kpf_data['geom']:
        kpf_data['actor_geom_map'][geomval['id1']].append(geomval)
    for activityval in kpf_data['activities']:
        for actor in activityval['actors']:
            kpf_data['actor_activity_map'][actor['id1']].append(activityval)

    return kpf_data


def convert(kpf_data: KPFData) -> Tuple[types.DIVEAnnotationSchema, dict]:
    """Convert kpf data to pre-validated DIVE Json"""
    tracks: Dict[str, dict] = {}
    groups: Dict[str, dict] = {}
    attribute_definitions = {
        'srcStatus': models.Attribute(
            **{
                'belongs': 'track',
                'datatype': 'text',
                'name': 'srcStatus',
                'key': 'srcStatus',
            }
        ).dict(),
        'activityIds': models.Attribute(
            **{
                'belongs': 'track',
                'datatype': 'text',
                'name': 'activityIds',
                'key': 'activityIds',
            }
        ).dict(),
    }

    for actorId in kpf_data['actor_geom_map'].keys():
        activities = kpf_data['actor_activity_map'][actorId]
        actorType = kpf_data['actor_type_map'][actorId]
        attributes = {
            'activityIds': ' '.join([str(a['id2']) for a in activities]),
            'srcStatus': ' '.join([str(a['src_status']) for a in activities]),
        }
        allActorInstances = [
            actor for act in activities for actor in act['actors'] if actor['id1'] == actorId
        ]
        allRanges = [
            t for actor in allActorInstances for ts in actor['timespan'] for t in ts['tsr0']
        ]
        begin = min(allRanges)
        end = max(allRanges)
        features: List[dict] = []
        confidencePairs = list(actorType['cset3'].items())
        for geom in kpf_data['actor_geom_map'][actorId]:
            if geom['keyframe']:
                features.append(
                    {
                        'frame': geom['ts0'],
                        'attributes': {},
                        'bounds': [float(a) for a in geom['g0'].split(' ')],
                        'interpolate': True,
                    }
                )
        tracks[str(actorType['id1'])] = models.Track(
            **{
                'attributes': attributes,
                'id': actorType['id1'],
                'begin': begin,
                'end': end,
                'confidencePairs': confidencePairs,
                'features': sorted(features, key=lambda f: f['frame']),
            }
        ).dict(exclude_none=True)

        for activity in kpf_data['activities']:
            all_ranges = list(flatten([a['tsr0'] for a in activity['timespan']]))
            confidence_pairs = list(activity['act2'].items())
            begin = min(all_ranges)
            end = max(all_ranges)
            members: Dict[int, dict] = {}
            for actor in activity['actors']:
                members[actor['id1']] = {
                    'ranges': [a['tsr0'] for a in actor['timespan']],
                }
            groups[str(activity['id2'])] = models.Group(
                begin=begin,
                end=end,
                members=members,
                id=activity['id2'],
                confidencePairs=confidence_pairs,
                attributes={},
            ).dict(exclude_none=True)

    return {
        'tracks': dict(sorted(tracks.items())),
        'groups': dict(sorted(groups.items())),
        'version': constants.AnnotationsCurrentVersion,
    }, models.MetadataMutable(attributes=attribute_definitions).dict(exclude_none=True)
